Add AlertManagerConfiguration CRD & collect instances into config

grdryn opened the pull request:
This is an implementation of a subset of the prometheus-operator
proposal which can be found here:
https://docs.google.com/document/d/1aVVttvocop8zNezwrNFbHh_HKRW_fhJUZqhmrr8fGK0/edit#

### Limitations & deviations from the above proposal:
  * This implementation doesn't enforce the namespace matcher on the
    first level Route in a CR, as described in the above
    proposal. This is because this operator only watches its own
    namespace.
  * The receiver types don't expose the ability to set HTTPConfig/TLSConfig
    as described in the proposal. I don't believe we need that immediately, but
    happy to add it if necessary.
  * The PagerDuty config in the CRD doesn't allow specification of "service_key".
    This field would have been mutually exclusive with the routing_key field (former
    is used with PagerDuty "Events v1" API, latter with "Events v2". Given that
    hardcoded config already uses routing_key/v2, I hope we don't need to cater
    for the v1 way (and the possibility for someone to accidentally set both fields).

grdryn commented:
/cc @david-martin @jewzaam 

While not complete, I think this has pretty much all the dots joined together, so I'd really appreciate a review if you have some time. :slightly_smiling_face:

jewzaam commented:
Now that https://github.com/openshift/configure-alertmanager-operator/pull/76 is merged it would be great to rebase this guy. Guess because of the merge commit it needs explicit rebase?.  @grdryn after rebase would you say this is ready for a first pass of review or should we wait for more of the todo stuff to land?

grdryn commented:
@jewzaam I've rebased now, updated the description, and removed the WIP tag. From my point of view, this is ready for review now

I'm especially interested to hear if you think the limitations/deviations outlined in the description are acceptable (some of these have TODO comments in the code (the HTTP/TLS config stuff), but I don't currently intend to do anything around them unless it's necessary); and also if you think any additional tests are needed anywhere.

I'm also not 100% sure what happens once this PR eventually gets merged. Is there somewhere that a CSV/operator bundle is stored for this operator that I should update? It will need to deploy the new CRD & updated Role, along with the new image version.

grdryn commented on pkg/apis/alertmanager/v1alpha1/alertmanagerconfiguration_types.go:
```diff
@@ -0,0 +1,551 @@
+		Details:        details,
+		Severity:       p.Severity,
+
+		// TODO: are these properties used?
+		Class:     "",
+		Component: "",
+		Group:     "",
```
The reason this is a TODO is that these fields weren't in the prometheus-operator proposal, and they weren't documented in the Alertmanager docs. I've created a PR to those docs to get them included.

I'll update this to make them available in the CRD, so there's one less TODO :slightly_smiling_face:

grdryn commented:
/lint

grdryn commented:
/lint

jewzaam commented:
/hold until we address configuration
I haven't read through the PR yet but assuming it aligns with the design we'd add to managed-cluster-config whatever CRs we need for OSD alert configuration to match current state.

grdryn commented:
@jewzaam that's an important point that I forgot to mention: this currently doesn't change how the existing OSD routes and receivers are created: those are still handled in the same way, just _additional_ config can be included with AlertManagerConfiguration CRs. If you'd like, I can update this to pull out that pre-existing config to CRs also?

grdryn commented:
Would make targets to generate the CRDs & deepcopy code be useful here? It's actually something I meant to add, but I wasn't sure which .mk file to add it to. It could be a pre-requisite target to `gobuild` or something.

For reference, the commands are:
```
operator-sdk generate crds
operator-sdk generate k8s
```

If you'd like, I can add the cli parts of the operator-sdk as a tools dependency, so that people (or automation) doesn't need to worry about having the operator-sdk on the path (or having the correct version).

cblecker commented:
/cc

grdryn commented:
I've added a new change here that converts the pre-existing hardcoded config to CRs. I'm happy to remove it from this pull request again if you'd prefer to add it as a follow-up change or something.

Here's the diff of the secret on a cluster before this version of the operator, and after deploying this version and the two CRs (with secret values redacted):

```diff
--- before.yaml	2020-06-05 18:13:07.982716233 +0100
+++ after.yaml	2020-06-05 18:52:24.705646064 +0100
@@ -6,84 +6,89 @@
   group_by:
   - job
   routes:
-  - receiver: "null"
+  - receiver: openshift-monitoring-osd-dms-am-configuration-osd-watchdog
+    match:
+      alertname: Watchdog
+    continue: true
+    repeat_interval: 5m
+  - receiver: openshift-monitoring-osd-pd-am-configuration-osd-pagerduty
     group_by:
     - alertname
     - severity
     continue: true
     routes:
-    - receiver: "null"
+    - receiver: openshift-monitoring-osd-pd-am-configuration-osd-null
       match:
         alertname: KubeQuotaExceeded
-    - receiver: "null"
+    - receiver: openshift-monitoring-osd-pd-am-configuration-osd-null
       match:
         alertname: UsingDeprecatedAPIAppsV1Beta1
-    - receiver: "null"
+    - receiver: openshift-monitoring-osd-pd-am-configuration-osd-null
       match:
         alertname: UsingDeprecatedAPIAppsV1Beta2
-    - receiver: "null"
+    - receiver: openshift-monitoring-osd-pd-am-configuration-osd-null
       match:
         alertname: UsingDeprecatedAPIExtensionsV1Beta1
-    - receiver: "null"
+    - receiver: openshift-monitoring-osd-pd-am-configuration-osd-null
       match:
         alertname: CPUThrottlingHigh
         container: registry-server
-    - receiver: "null"
+    - receiver: openshift-monitoring-osd-pd-am-configuration-osd-null
       match:
         alertname: CPUThrottlingHigh
         container: configmap-registry-server
-    - receiver: "null"
+    - receiver: openshift-monitoring-osd-pd-am-configuration-osd-null
       match:
         alertname: NodeFilesystemSpaceFillingUp
         severity: warning
-    - receiver: "null"
+    - receiver: openshift-monitoring-osd-pd-am-configuration-osd-null
       match:
         namespace: openshift-customer-monitoring
-    - receiver: "null"
+    - receiver: openshift-monitoring-osd-pd-am-configuration-osd-null
       match:
         namespace: openshift-operators
-    - receiver: "null"
+    - receiver: openshift-monitoring-osd-pd-am-configuration-osd-null
       match:
         alertname: SLAUptimeSRE
-    - receiver: "null"
+    - receiver: openshift-monitoring-osd-pd-am-configuration-osd-null
       match:
         alertname: CustomResourceDetected
-    - receiver: "null"
+    - receiver: openshift-monitoring-osd-pd-am-configuration-osd-null
       match:
         alertname: ImagePruningDisabled
-    - receiver: "null"
+    - receiver: openshift-monitoring-osd-pd-am-configuration-osd-null
       match:
         severity: info
-    - receiver: make-it-warning
+    - receiver: openshift-monitoring-osd-pd-am-configuration-osd-make-it-warning
       match:
         alertname: KubeAPILatencyHigh
         severity: critical
-    - receiver: pagerduty
+    - receiver: openshift-monitoring-osd-pd-am-configuration-osd-pagerduty
       match_re:
         exported_namespace: ^default$|^kube.*|^openshift.*|^redhat-.*
-    - receiver: pagerduty
+    - receiver: openshift-monitoring-osd-pd-am-configuration-osd-pagerduty
       match:
         exported_namespace: ""
       match_re:
         namespace: ^default$|^kube.*|^openshift.*|^redhat-.*
-    - receiver: pagerduty
+    - receiver: openshift-monitoring-osd-pd-am-configuration-osd-pagerduty
       match:
         job: fluentd
-    - receiver: pagerduty
+    - receiver: openshift-monitoring-osd-pd-am-configuration-osd-pagerduty
       match:
         alertname: FluentdNodeDown
-    - receiver: pagerduty
+    - receiver: openshift-monitoring-osd-pd-am-configuration-osd-pagerduty
       match:
         cluster: elasticsearch
-  - receiver: watchdog
-    match:
-      alertname: Watchdog
-    repeat_interval: 5m
   group_wait: 30s
   group_interval: 5m
   repeat_interval: 12h
 receivers:
-- name: pagerduty
+- name: openshift-monitoring-osd-dms-am-configuration-osd-watchdog
+  webhook_configs:
+  - send_resolved: true
+    url: REDACTED
+- name: openshift-monitoring-osd-pd-am-configuration-osd-pagerduty
   pagerduty_configs:
   - send_resolved: true
     routing_key: REDACTED
@@ -93,8 +98,8 @@
       component: '{{ .CommonLabels.alertname }}'
       group: '{{ .CommonLabels.alertname }}'
       link: '{{ if .CommonAnnotations.link }}{{ .CommonAnnotations.link }}{{ else
-        }}https://github.com/openshift/ops-sop/tree/master/v4/alerts/{{ .CommonLabels.alertname
-        }}.md{{ end }}'
+        }}https://github.com/openshift/ops-sop/tree/master/v4/alerts/{{ .CommonLabels.alertname}}.md{{
+        end }}'
       link2: '{{ if .CommonAnnotations.runbook }}{{ .CommonAnnotations.runbook }}{{
         else }}{{ end }}'
       num_firing: '{{ .Alerts.Firing | len }}'
@@ -102,7 +107,7 @@
       resolved: '{{ template "pagerduty.default.instances" .Alerts.Resolved }}'
     severity: '{{ if .CommonLabels.severity }}{{ .CommonLabels.severity | toLower
       }}{{ else }}critical{{ end }}'
-- name: make-it-warning
+- name: openshift-monitoring-osd-pd-am-configuration-osd-make-it-warning
   pagerduty_configs:
   - send_resolved: true
     routing_key: REDACTED
@@ -120,9 +125,6 @@
       num_resolved: '{{ .Alerts.Resolved | len }}'
       resolved: '{{ template "pagerduty.default.instances" .Alerts.Resolved }}'
     severity: warning
-- name: watchdog
-  webhook_configs:
-  - send_resolved: true
-    url: REDACTED
+- name: openshift-monitoring-osd-pd-am-configuration-osd-null
 - name: "null"
 templates: []
```

grdryn commented:
Rebased & conflict resolved.

RiRa12621 commented:
/assign @RiRa12621

RiRa12621 commented:
/cc @lilic @metalmatze @brancz 
For extra eyes

brancz commented:
Frankly I'm a little confused why this wasn't even attempted to be contributed to the prometheus operator directly, a project that we almost entirely control. This implementation is even based on the proposal written for it. Can someone explain?

RiRa12621 commented:
@brancz @simonpasquier and @grdryn are going to try to lift this into prometheus-operator

/hold
until 1st checkpoint in two weeks time, July 15th 2020

lilic commented:
> @brancz @simonpasquier and @grdryn are going to try to lift this into prometheus-operator

FYI in case someone didn't mention this before there is a bunch of work done for this in upstream already so you can merge it with this one, looking forward to the contributions there! 🎉  :) https://github.com/coreos/prometheus-operator/pull/3265

brancz commented:
The agreement was that from the 1st of July until today the RHMI team was supposed to attempt to complete that PR.

I have to be honest that I'm a little disappointed, that even though putting 2 weeks of work into finishing the above PR was the agreed upon outcome of an explicit meeting on the topic with meeting notes on this agreement, we still didn't put a single day of work into this.

RiRa12621 commented:
not sure how we're gonna go forward from here with this.
/assign @jeremyeder 
needs architect support on how to move ahead

RiRa12621 commented:
/unassign @RiRa12621

jeremyeder commented:
Agreed, this is on pause until we get to a decision either way.

grdryn commented:
Apologies, I should have closed this pull request already. We won't need it now, so I won't be keeping it up-to-date or following up to get it merged.

Even though we did agree that us RHMI folks would try to progress the changes in the prometheus-operator in those 2 weeks, within a day that plan was changed due to other priorities and demands on our time (sorry if that message didn't propagate to everyone, I assumed it had). It's in our minds to try to find the time to do it in the short to medium term...if we do find that time we'll be in touch! :slightly_smiling_face:
